### PR TITLE
Cryptokitties

### PR DESCRIPTION
**Pull request by dorin-iancu**
Cryptokitties implementation, split into 3 smart contracts:
1. KittyOwnership - The main contract. It holds all the information regarding who owns which kitty, handles breeding logic, and it performs many other checks. It also implements the ERC721 interface.
2. KittyGeneticAlg - A genetic algorithm mock. This can be easily extended in the future, but I made it as simple as possible for the sake of the example.
3. KittyAuction - Handles the kitty auctioning, where players may either sell their kitties or allow other players to have one of their kitties mate with theirs.

### CODE REVIEW

File: contracts/examples/crypto-kitties/kitty-auction/src/lib.rs [hunk 1]
```diff
@@ -0,0 +1,334 @@
...
+			AsyncCallResult::Err(_) => {
+				// this can only fail if the kitty_ownership contract address is invalid
+				// nothing to revert in case of error
+			},
+		}
+	}
+
+	// private
+
+	fn _create_auction(
```
Comment (andrei-marinica) [inline contracts/examples/crypto-kitties/kitty-auction/src/lib.rs:275]: no need for leading underscores anymore

File: contracts/examples/crypto-kitties/kitty-genetic-alg/src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+#[elrond_wasm_derive::contract(KittyGeneticAlgImpl)]
+pub trait KittyGeneticAlg {
+	#[init]
+	fn init(&self) {}
+
+	// endpoints
+
+	#[endpoint(generateKittyGenes)]
+	fn generate_kitty_genes(&self, matron: Kitty, sire: Kitty) -> SCResult<KittyGenes> {
+		let mut random = Random::new(*self.get_block_random_seed());
```
Comment (andrei-marinica) [inline contracts/examples/crypto-kitties/kitty-genetic-alg/src/lib.rs:17]: Isn't the block random seed already random enough?

Comment (andrei-marinica) [inline contracts/examples/crypto-kitties/kitty-genetic-alg/src/lib.rs:17]: Isn't the block random seed random enough?
Oh, I see, we need more than 1 number.
If we just need 2 random numbers 1-99, I think we can split the block random seed. (Thinking now, if we shouldn't provide something from the framework that does that).

File: contracts/examples/crypto-kitties/kitty-genetic-alg/src/lib.rs [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+		let fur_color_percentage = 1 + random.next_u8() % 99; // val in [1, 100)
+		let matron_fur_color = matron.get_fur_color();
+		let sire_fur_color = sire.get_fur_color();
+		let kitty_fur_color = matron_fur_color.mix_with(
+			&sire_fur_color,
+			fur_color_percentage,
+			100 - fur_color_percentage,
+		);
+
+		let eye_color_percentage = 1 + random.next_u8() % 99; // val in [1, 100)
```
Comment (andrei-marinica) [inline contracts/examples/crypto-kitties/kitty-genetic-alg/src/lib.rs:28]: Shouldn't we also somehow include the tx nonce into the randomness? Calling generateKittyGenes twice in the same block could produce 2 identical cats. Is that ok?

**Review by andrei-marinica — APPROVED** [review]

File: contracts/examples/crypto-kitties/kitty-auction/src/lib.rs (hunk 1 above)
Comment (dorin-iancu) [reply contracts/examples/crypto-kitties/kitty-auction/src/lib.rs:275]: I prefer having underscores for private functions, but I'm willing to change that if it's not a good convention.

File: contracts/examples/crypto-kitties/kitty-genetic-alg/src/lib.rs (hunk 2 above)
Comment (dorin-iancu) [reply contracts/examples/crypto-kitties/kitty-genetic-alg/src/lib.rs:17]: In the current implementation, yes, I could simply take the block random seed and split it, but I prefer having it this way for easier extension later.

**Review by sasurobert — CHANGES_REQUESTED** [review]

**Review by sasurobert — APPROVED** [review]

**Pull request merged**